// @figma/code-connect/dist/common/external.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.textContent = exports.className = exports.nestedProps = exports.children = exports.instance = exports.string = exports.enum = exports.boolean = exports.connect = void 0;
const React = __importStar(require("react"));
function connectType(_component, _figmaNodeUrl, _meta) { }
exports.connect = connectType;
function booleanType(_figmaPropName, _valueMapping) {
    if (_valueMapping) {
        return enumType(_figmaPropName, _valueMapping);
    }
    return true;
}
exports.boolean = booleanType;
function enumType(_figmaPropName, _valueMapping) {
    return Object.values(_valueMapping)[0];
}
exports.enum = enumType;
function stringType(_figmaPropName) {
    return '';
}
exports.string = stringType;
function instanceType(_figmaPropName) {
    return React.createElement('div');
}
exports.instance = instanceType;
function childrenType(_layers) {
    return React.createElement('div');
}
exports.children = childrenType;
function nestedPropsType(_layer, props) {
    return props;
}
exports.nestedProps = nestedPropsType;
function classNameType(_className) {
    return '';
}
exports.className = classNameType;
function textContentType(_layer) {
    return '';
}
exports.textContent = textContentType;
