// @figma/code-connect/dist/connect/create.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCodeConnectFromUrl = exports.normalizeComponentName = void 0;
const helpers_1 = require("./helpers");
const axios_1 = __importStar(require("axios"));
const fs_1 = __importDefault(require("fs"));
const figma_rest_api_1 = require("./figma_rest_api");
const logging_1 = require("../common/logging");
const parser_executables_1 = require("./parser_executables");
const create_1 = require("../react/create");
const parser_executable_types_1 = require("./parser_executable_types");
const zod_validation_error_1 = require("zod-validation-error");
function normalizeComponentName(name) {
    // Convert the string to PascalCase and ensure first character is not a digit
    return name
        .replace(/[^a-zA-Z0-9]/g, ' ')
        .split(' ')
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join('')
        .replace(/^[0-9]/, '_$&');
}
exports.normalizeComponentName = normalizeComponentName;
async function createCodeConnectFromUrl({ accessToken, figmaNodeUrl, outFile, outDir, projectInfo, }) {
    try {
        const fileKey = (0, helpers_1.parseFileKey)(figmaNodeUrl);
        const nodeIds = (0, helpers_1.parseNodeIds)([figmaNodeUrl]);
        const apiUrl = (0, figma_rest_api_1.getApiUrl)(figmaNodeUrl ?? '') + `/code_connect/${fileKey}/cli_data?ids=${nodeIds.join(',')}`;
        if (nodeIds.length === 0) {
            (0, logging_1.exitWithError)(`Invalid figma node URL: the provided url "${figmaNodeUrl}" does not contain a node-id`);
        }
        else if (nodeIds.length > 1) {
            (0, logging_1.exitWithError)(`Invalid figma node URL: the provided url "${figmaNodeUrl}" contains more than one node-id`);
        }
        logging_1.logger.info('Fetching component information from Figma...');
        const response = process.env.CODE_CONNECT_MOCK_CREATE_API_RESPONSE
            ? {
                status: 200,
                data: JSON.parse(fs_1.default.readFileSync(process.env.CODE_CONNECT_MOCK_CREATE_API_RESPONSE, 'utf-8')),
            }
            : await axios_1.default.get(apiUrl, {
                headers: (0, figma_rest_api_1.getHeaders)(accessToken),
            });
        if (response.status === 200) {
            logging_1.logger.info('Parsing response');
            const component = (0, helpers_1.findComponentsInDocument)(response.data.document, nodeIds)[0];
            const normalizedName = normalizeComponentName(component.name);
            const payload = {
                mode: 'CREATE',
                destinationDir: outDir ?? process.env.INIT_CWD ?? process.cwd(),
                destinationFile: outFile,
                component: {
                    figmaNodeUrl,
                    id: component.id,
                    name: component.name,
                    normalizedName,
                    type: component.type,
                    componentPropertyDefinitions: component.componentPropertyDefinitions,
                },
                config: projectInfo.config,
            };
            logging_1.logger.info('Generating Code Connect files...');
            let result;
            if (projectInfo.config.parser === 'react') {
                result = await (0, create_1.createReactCodeConnect)(payload);
            }
            else {
                try {
                    const stdout = await (0, parser_executables_1.callParser)(
                    // We use `as` because the React parser makes the types difficult
                    // TODO remove once React is an executable parser
                    projectInfo.config, payload, projectInfo.absPath);
                    result = parser_executable_types_1.CreateResponsePayload.parse(stdout);
                }
                catch (e) {
                    throw (0, zod_validation_error_1.fromError)(e);
                }
            }
            const { hasErrors } = (0, parser_executables_1.handleMessages)(result.messages);
            if (hasErrors) {
                (0, logging_1.exitWithError)('Errors encountered calling parser, exiting');
            }
            logging_1.logger.info('Code Connect files generated successfully:');
            result.createdFiles.forEach((file) => {
                logging_1.logger.info(`${file.filePath}`);
            });
        }
        else {
            logging_1.logger.error(`Failed to get node information from Figma with status: ${response.status}`);
            logging_1.logger.debug('Failed to get node information from Figma with Body:', response.data);
        }
    }
    catch (err) {
        if ((0, axios_1.isAxiosError)(err)) {
            if (err.response) {
                logging_1.logger.error(`Failed to get node data from Figma (${err.code}): ${err.response?.status} ${err.response?.data?.err ?? err.response?.data?.message}`);
            }
            else {
                logging_1.logger.error(`Failed to get node data from Figma: ${err.message}`);
            }
            logging_1.logger.debug(JSON.stringify(err.response?.data));
        }
        else {
            logging_1.logger.error(`Failed to create: ${err}`);
        }
        (0, helpers_1.exitWithFeedbackMessage)(1);
    }
}
exports.createCodeConnectFromUrl = createCodeConnectFromUrl;
